Performance > Duplicate queries for @mentions when rendering activity feed - Fixed

mahedihasannoman opened the pull request:
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Prevented initiating new activity object(BP_Activity_Activity) if the activity object is already available in the global activity template.

Fixes #3070 .

### How to test the changes in this Pull Request:

1. Create an activity post with multiple @mentions
2. Capture the queries for the activity item where has multiple @mentions when rendering.
3. See no duplicate queries

### Proof Screenshots or Video

https://www.loom.com/share/0a90b9669792476c8a0fa3245177a30b

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

chetansatasiya commented on src/bp-activity/bp-activity-functions.php:
```diff
@@ -3541,10 +3541,22 @@ function bp_activity_delete_children( $activity_id, $comment_id ) {
+		 * If available then take the activity object from global template.
+		 * Else create a new instance.
+		 */
+		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $activity_id ) {
```
```suggestion
		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $activity_id ) {
```

chetansatasiya commented on src/bp-forums/activity.php:
```diff
@@ -416,14 +416,30 @@ public function before_activity_content( $content, $activity ) {
+			 * If available then take the activity object from global template.
+			 * Else create a new instance.
+			 */
+			if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $activity_id ) {
```
```suggestion
			if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $activity_id ) {
```

chetansatasiya commented on src/bp-moderation/classes/class-bp-moderation-activity.php:
```diff
@@ -97,7 +97,19 @@ public static function get_permalink( $activity_id ) {
+		 * If available then take the activity object from global template.
+		 * Else create a new instance.
+		 */
+		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $activity_id ) {
```
```suggestion
		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $activity_id ) {
```

chetansatasiya commented on src/bp-moderation/classes/class-bp-moderation-activity.php:
```diff
@@ -178,7 +190,18 @@ public function restrict_single_item( $restrict, $activity ) {
+		 * If available then take the activity object from global template.
+		 * Else create a new instance.
+		 */
+		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $item_id ) {
```
```suggestion
		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $item_id ) {
```

chetansatasiya commented on src/bp-moderation/classes/class-bp-moderation-activity.php:
```diff
@@ -328,11 +351,23 @@ public function update_button_sub_items( $item_id ) {
+		 * If available then take the activity object from global template.
+		 * Else create a new instance.
+		 */
+		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $item_id ) {
```
```suggestion
		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $item_id ) {
```

chetansatasiya reviewed (changes requested):
@mahedihasannoman Use the strict comparison in all the places in condition.

chetansatasiya reviewed (changes requested):
@mahedihasannoman I debugged the issue and there is nothing wrong with these functions you made a change on this PR.

The issue is here https://github.com/buddyboss/buddyboss-platform/blob/release/src/bp-members/bp-members-functions.php#L307 as it's making a lot of queries to DB.

We can add the cache on this function like this.

```

function bp_core_get_userid_from_nickname( $nickname ) {
	if ( empty( $nickname ) ) {
		return false;
	}

	static $cache = array();

	$cache_key = 'bp_core_get_userid_from_nickname' . $nickname;
	if ( ! isset( $cache[ $cache_key ] ) ) {
		$cache[ $cache_key ] = get_users(
			array(
				'meta_key'    => 'nickname',
				'meta_value'  => $nickname,
				'number'      => 1,
				'count_total' => false,
			)
		);

		$user = $cache[ $cache_key ] ? $cache[ $cache_key ][0] : null;
	} else {
		$user = $cache[ $cache_key ] ? $cache[ $cache_key ][0] : null;
	}

	return apply_filters( 'bp_core_get_userid_from_nickname', ! empty( $user->ID ) ? $user->ID : null, $nickname );
}

```

mahedihasannoman commented:
Hi @chetansatasiya , I have added cache on that function as you suggested and it is working fine. Also I have removed all of my previous code and conditions. Thanks.

sumitbuddyboss commented:
Hello @mahedihasannoman 

The related issue is resolved, Now no more duplicate queries for @mention when the activity feed is rendering.

**Video proof:**
https://www.loom.com/share/9c8d80561a6b4892b5239cd41a163e6d

**Testcase:**
https://docs.google.com/spreadsheets/d/1v7bvYc-n-VZYQu8CcAcc9jLmHFduAQmxz_0n8lhjw7o/edit?pli=1#gid=1451538933

Ready for release 🚀 

Thank you

sumitbuddyboss commented:
Hello @mahedihasannoman 

I have tested in a qa-approved branch and ready for release.

Test staus: PASS

Ready for release 🚀 

Thank you
